### ISSUE TITLE
Probability threshold too aggressive

### ISSUE THREAD

**Issue by faulk-lab** [nanoporetech/modkit#3]
Me again,

Today I'm calculating global methylation on a 30X coverage primate genome blood sample, using modkit defaults.
`~/Desktop/modkit0.1.3/modkit pileup -t 32 --cpg --ref panTro6.fa --combine-strands total-chimp.modmapped.bam total-chimp.modmapped.bam.modkit.bed`. The automatically determined threshold seem to be too aggressive about filtering. On this data set, modkit default chooses `threshold 0.9902344`.

When I use defaults and summarize the 5mc with awk, I get 88.9% global 5mC, which is rather high. The same calculation from `modbam2bed` using defaults comes in at 77.8%. I re-ran modkit with `modkit --filter-threshold 0.66`, I get 79.7%. That's more reasonable. Primate blood methylation really shouldn't be higher than 80%.

Taking a look at the total counts of modified CpG calls I see this:
264,978,433 5mc calls `modkit` default
28,821,284 lines

276,737,102 5mc calls `modkit --filter-threshold 0.66`
29,010,994 lines

276,147,783 5mc calls `modbam2bed` default
29,395,297 lines

So, the total number of calls missing is just a small fraction ~12million calls missing (or just 1% of them) at the 99% threshold level. That seems like it shouldn't be a problem. So how can my methylation be jumping from 79.7% to 88.9%??

I tried increasing the sampling fraction and that didn't help. The number of lines is nearly the same so modkit is calling all the reference CpGs. I believe the threshold sampling algorithm is too strict in kicking out modified bases. Of course I could just manually set my own threshold but that defeats the purpose of using modkit in the first place.

Modkit sample-probs on the bam file gives me this output which I don't know how to interpret.
~/Desktop/modkit0.1.3/modkit sample-probs -t 32 total-chimp.modmapped.bam> sampling 10042 reads from BAM to estimate probability distribution
▹▸▹▹▹ [00:02:05] 1322327 probabilities sampled                                                                                                           > sampled 1322327 probabilities
q       p
0.10    0.99
0.50    1.00
0.90    1.00

**Comment by ArtRand** [comment]
Hey @faulk-lab, 

The `modkit sample-probs` outputs descriptive statistics of the base modification call probability distribution. The default is to output the 10th, 50th, and 90th percentiles, which is what you have. Note that it's not the probability of being methylated, it's the confidence of the model in the call, either methylated or canonical.

Would you mind telling me which model you're using? I'm assuming one of the 5mC/C models, 0.99 at the 10th percentile seems quite high. Could you tell me what `modkit sample-probs -p 0.01,0.05,0.1` produces? `modkit summary` without filtering the calls may also help figure out what's going on. Basically it looks like the canonical calls are all low-confidence.

**Comment by faulk-lab** [comment]
The model called was `dna_r10.4.1_e8.2_400bps_modbases_5mc_cg_sup.cfg` through Guppy 6.3.9. 

`modkit sample-probs -p 0.01,0.05,0.1` outputs the following:
> sampled 1322327 probabilities
q       p
0.01    0.72
0.05    0.97
0.10    0.99

`modkit summary --no-filtering` gives me this, which is frankly, pretty much on target.
mod_bases       C
count_reads_C   6634378
C_calls_modified_m      305553221
C_frac_modified_m       0.7801294870064018
C_filtered_modified_m   0
C_calls_unmodified      86116657
C_frac_unmodified       0.21987051299359814
C_filtered_unmodified   0
C_total_mod_calls       391669878
C_total_filtered_mod_calls      0
total_reads_used        6634378

**Comment by ArtRand** [comment]
The issue could be that because the 5mC/C models tend to emit very high confidence calls, the binning of the probabilities is causing the filtering rule (filter out <= threshold) to remove too many calls. What the `sample-probs` is telling me is that modkit is requiring the call be in the top two bins. 

I've adjusted the code to make the rule remove calls less than the threshold value (but keep calls equal to the threshold value). I've uploaded a release candidate to [0.1.4-rc1](https://github.com/nanoporetech/modkit/releases/tag/v0.1.4-rc1). I'm going to run some tests early this week, but would also greatly appreciate any additional feedback you have from the analysis of your sample. Thanks in advance.

**Comment by faulk-lab** [comment]
This seems like a better filter rule in 0.1.4-rc1. It brings my chimp down to 82.3% global 5mC with a filter threshold 0.9902344. That's still higher than modbam2bed's 77.8% but more reasonable. I ran a few other primates as well, with baboon at 82.3% (up from 75.8% in modbam2bed) and rhesus at 77.0% (up from 71.8% in modbam2bed). I'm comfortable with the biological plausibility of these numbers now. I think this is a good threshold setting. It becomes a difficult question to answer because nobody has ever measured global methylation quite like this. To really determine the accuracy of the thresholds you would need controls. Fortunately I have them.

I created 100% methylated mouse DNA and ran it on kit14 with the same basecalling models as the primates. Version 0.1.4-rc1 reports the 100% as 0.99354 5mCpG modified, with a filter threshold 0.9824219. This is a great improvement from modbam2bed using defaults that came in at 94.8%. I believe modbam2bed's thresholds were too low.

I created 0% methylated mouse DNA and ran it. Version 0.1.4-rc1 reports it as 0.0108157 5mCpG modified, with a filter threshold 0.9628906. (It was pretty garbage DNA after the WGA kit so I'm not surprised at the low filter threshold). That's also an improvement from the 1.48% methylation reported by modbam2bed.

I think you've got reasonable settings in this version. I'll be cranking the rest of my data through it and I'll let you know if I run into anything else weird. I super appreciate your responsiveness! Fantastic program. I'm looking forward to making it the key step in my pipeline!
